File: solutions_python/problems/arrays/SubArraySort.py


# import the module 
import sys 
def out_of_order(arr,idx):
    
    val = arr[idx]
    if idx==0 :
        return val>arr[1]
    if idx==len(arr)-1:
        return val <arr[len(arr)-2]
    return val > arr[idx+1] or val < arr[idx-1]
def sub_array_sort(arr):
    smallestVal = float('-inf')
    largestVal= float('inf')
    for i in range(len(arr)):
        if out_of_order(arr, i):
            largestVal= max(largestVal, arr[i])
            smallestVal= min(smallestVal, arr[i])
            
    if smallestVal==float('-inf'):
        print("-1, -1")
        return        
    leftIdx= 0
    rightIdx= len(arr)-1
    while arr[leftIdx]<=smallestVal:
        leftIdx +=1
    while arr[rightIdx]>=largestVal:
        rightIdx-=1
    print("-----LEFT IDX IS ",leftIdx, " RIGHT IDX IS", rightIdx  ) 
    
arr= [1, 2, 3, 4,5,6,7, 8 ,9, 10,11]
sub_array_sort(arr)           

File: solutions_java/problems/arrays/SubArraySort.java
package solutions_java.problems.arrays;

import java.util.Arrays;
import java.util.List;

public class SubArraySort {
     boolean outOfOrder(List<Integer> arr, int idx) {
        int val = arr.get(idx);
        if (idx == 0) {
            return val > arr.get(1);
        }
        if (idx == arr.size() - 1) {
            return val < arr.get(idx - 1);
        }

        return val > arr.get(idx + 1) || val < arr.get(idx - 1);
    }

     void subarraySort(List<Integer> arr) {
        int smallestVal = 1000;
        int largestVal = -1000;
        for (int i = 0; i < arr.size(); i++) {
            int val = arr.get(i);
            if (outOfOrder(arr, i)) {
                smallestVal = Math.min(smallestVal, val);
                largestVal = Math.max(largestVal, val);
            }

        }
        if (smallestVal == Integer.MAX_VALUE) {
            System.out.println("LEFT INDEX IS -1, RIGHT INDEX IS -1");
        }
        int left = 0;
        int right = arr.size() - 1;

        while (arr.get(left) <= smallestVal) {

            left++;
        }

        while (arr.get(right) >= largestVal) {
            right--;
        }

        System.out.println("LEFT IDX IS " + left + " RIGHT IDX IS " + right);

    }

    public static void main(String[] args) {
        List<Integer> arr = Arrays.asList(1, 2, 3, 4, 5, 8, 6, 7, 9, 10, 11);
        new SubArraySort().subarraySort(arr);
    }

}
